# K-Fold Cross Validation
**Cross validation is using for prevent overfitting. Here we are using K-Fold Cross Validation with Support Vector Machines.**

Cross validation (CV) is one of the technique used to test the effectiveness of a machine learning models, it is also a re-sampling procedure used to evaluate a model if we have a limited data. K-Fold CV is where a given data set is split into a K number of sections/folds where each fold is used as a testing set at some point. .

In [0]:
import numpy as np                                    # import numpy for  arrays and matrices operation.
from sklearn import model_selection                   # importing sklearn model selection for Split arrays or matrices into random train and test subsets
from sklearn.model_selection import cross_val_score   # importing cross_val_score for cross validation score measurement and metrices.
from sklearn.model_selection import train_test_split  # for split arrays or matrices into random train and test subsets
from sklearn import datasets                          # imporrt sklearn datasets for loading prebuild datasets.
from sklearn import svm                               # import sklearn svm module for analyze data used for classification

To disable the python future warings on packages and libraries we will use warning packages. It will keep the code clean. But its nor recommended.

In [0]:
import warnings
warnings.filterwarnings("ignore", category=FutureWarning)

**Here I am using prebuilt -  wine - dataset in sklearn.**

In [0]:
dataset = datasets.load_wine()  # loading the wine dataset as numpy array.
print(dataset)                  # pring the array values.

{'data': array([[1.423e+01, 1.710e+00, 2.430e+00, ..., 1.040e+00, 3.920e+00,
        1.065e+03],
       [1.320e+01, 1.780e+00, 2.140e+00, ..., 1.050e+00, 3.400e+00,
        1.050e+03],
       [1.316e+01, 2.360e+00, 2.670e+00, ..., 1.030e+00, 3.170e+00,
        1.185e+03],
       ...,
       [1.327e+01, 4.280e+00, 2.260e+00, ..., 5.900e-01, 1.560e+00,
        8.350e+02],
       [1.317e+01, 2.590e+00, 2.370e+00, ..., 6.000e-01, 1.620e+00,
        8.400e+02],
       [1.413e+01, 4.100e+00, 2.740e+00, ..., 6.100e-01, 1.600e+00,
        5.600e+02]]), 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2,
 

In [0]:
print(dataset.data.shape)     # return the shape of the dataset.
print('\n')
print(dataset.feature_names)  # return the name of the columns in the dataset.
print('\n')
print(dataset.target_names)   # return the values in the target column.
print('\n')
print(dataset.target)         # print the target set values.

(178, 13)


['alcohol', 'malic_acid', 'ash', 'alcalinity_of_ash', 'magnesium', 'total_phenols', 'flavanoids', 'nonflavanoid_phenols', 'proanthocyanins', 'color_intensity', 'hue', 'od280/od315_of_diluted_wines', 'proline']


['class_0' 'class_1' 'class_2']


[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2]


Dataset comprises of 13 features (alcohol, malic_acid, ash, alcalinity_of_ash, magnesium, total_phenols, flavanoids, nonflavanoid_phenols, proanthocyanins, color_intensity, hue, od280/od315_of_diluted_wines, proline) and type of wine cultivar. This data has three type of wine Class_0, Class_1, and Class_3. Here you can build a model to classify the type of wine. 

In [0]:
# Split the wine data into train/test data sets with 40% reserved for testing
X_train, X_test, y_train, y_test = model_selection.train_test_split(dataset.data, dataset.target, test_size=0.4, random_state=0)

# Build an SVC model for predicting wine classifications using training data
clf = svm.SVC(kernel='linear', C=1).fit(X_train, y_train)

# Now measure its performance with the test data
clf.score(X_test, y_test)   

0.9583333333333334

**For our model SVM classification model without cross validation, the classification accuracy is : 0.958 ie 95.83%.**

A single train/test split is made easy with the train_test_split function in the cross_validation library. We will split the dataset in to 60:40 propotion by keeping the 40% for test data. We  will assign more data for training to train our model well enough other wise having less amount of data for traning will leads to underfitting.

In [0]:
# We give cross_val_score a model, the entire data set and its "real" values, and the number of folds:
scores = cross_val_score(clf, dataset.data, dataset.target, cv=5)

# Print the accuracy for each fold:
print(scores)

# And the mean accuracy of all 5 folds:
print(scores.mean())

[0.94594595 0.88888889 0.97222222 1.         1.        ]
0.9614114114114114


K-Fold cross validation is just as easy; let's use a K of 5.

Here we are using scenario of 5-Fold cross validation(K=5). Here, the data set is split into 5 folds. In the first iteration, the first fold is used to test the model and the rest are used to train the model. In the second iteration, 2nd fold is used as the testing set while the rest serve as the training set. This process is repeated until each fold of the 5 folds have been used as the testing set

For our first model with fold value 5, the, the performance measure of each fold is given. **The overall average classification accuracy is 96.18%.** From the comparision its clear that the performance of the model is improved (for older model the score was 95.83%).

In [0]:
# Build an SVC model for predicting wine classifications using training data
clf = svm.SVC(kernel='poly', C=1).fit(X_train, y_train)

# Now measure its performance with the test data
clf.score(X_test, y_test)   

0.9722222222222222

Here when we change the kernael of the SVM to polynomial from linear with the same number of fold, **the classification performance is increased to 0.972**, which is better than the second model witth 5 folds and linear kernal.

*The linear and polynomial are simply different in case of making the hyperplane decision boundary between the classes.
The kernel functions are used to map the original dataset (linear/nonlinear ) into a higher dimensional space with view to making it linear dataset.
Usually linear and polynomial kernels are less time consuming.*